Microsoft.Toolkit.HighPerformance/Memory/Memory2D{T}.cs

```diff
@@ -0,0 +1,238 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.ComponentModel;
+using System.Diagnostics.Contracts;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Microsoft.Toolkit.HighPerformance.Extensions;
+
+namespace Microsoft.Toolkit.HighPerformance.Memory
+{
+    /// <summary>
+    /// <see cref="Memory2D{T}"/> represents a 2D region of arbitrary memory. It is to <see cref="Span2D{T}"/>
+    /// what <see cref="Memory{T}"/> is to <see cref="Span{T}"/>. For further details on how the internal layout
+    /// is structured, see the docs for <see cref="Span2D{T}"/>. The <see cref="Memory2D{T}"/> type can wrap arrays
+    /// of any rank, provided that a valid series of parameters for the target memory area(s) are provided.
+    /// </summary>
+    /// <typeparam name="T">The type of items in the current <see cref="Span2D{T}"/> instance.</typeparam>
+    public readonly struct Memory2D<T> : IEquatable<Memory2D<T>>
+    {
+        /// <summary>
+        /// The target <see cref="Array"/> instance, if present.
+        /// </summary>
+        private readonly Array? array;
+
+        /// <summary>
+        /// The initial offset within <see cref="array"/>.
+        /// </summary>
+        private readonly IntPtr offset;
+
+        /// <summary>
+        /// The height of the specified 2D region.
+        /// </summary>
+        private readonly int height;
+
+        /// <summary>
+        /// The width of the specified 2D region.
+        /// </summary>
+        private readonly int width;
+
+        /// <summary>
+        /// The pitch of the specified 2D region.
+        /// </summary>
+        private readonly int pitch;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Memory2D{T}"/> struct.
+        /// </summary>
+        /// <param name="array">The target array to wrap.</param>
+        /// <exception cref="ArrayTypeMismatchException">
+        /// Thrown when <paramref name="array"/> doesn't match <typeparamref name="T"/>.
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Memory2D(T[,]? array)
+        {
+            if (array is null)
+            {
+                this = default;
+
+                return;
+            }
+
+            if (!typeof(T).IsValueType && array.GetType() != typeof(T[]))
+            {
+                ThrowArrayTypeMismatchException();
```

Comment:
technically user types shouldn't throw this exception, but i think it is suitable here given it is trying to achieve parity with the system types. Added this note for anyone who comes along and has the same thought 🙂